main/src/vocab/nllb_vocab.rs

```diff
@@ -0,0 +1,238 @@
+        &self.values
+    }
+
+    fn indices(&self) -> &HashMap<i64, String> {
+        &self.indices
+    }
+
+    fn special_values(&self) -> &HashMap<String, i64> {
+        &self.special_values
+    }
+
+    fn special_indices(&self) -> &HashMap<i64, String> {
+        &self.special_indices
+    }
+
+    fn from_file<V: AsRef<Path>, S: AsRef<Path>>(
+        vocab: V,
+        special: Option<S>,
+    ) -> Result<Self, TokenizerError> {
+        let special_token_storage: Option<SpecialTokens> = special
+            .map(|p| {
+                File::open(&p)
+                    .context(IOSnafu { path: p.as_ref() })
+                    .and_then(|r| serde_json::from_reader(r).context(JsonDeserializeSnafu))
+            })
+            .transpose()?;
+
+        let mut tokenizer: Tokenizer = File::open(&vocab)
+            .context(IOSnafu {
+                path: vocab.as_ref(),
+            })
+            .map(BufReader::new)
+            .and_then(|r| serde_json::from_reader(r).context(JsonDeserializeSnafu))?;
+
+        let mut special_values = HashMap::with_capacity(FAIRSEQ_LANGUAGE_CODES.len() + 10);
+
+        let values = &mut tokenizer.model.vocab;
+        for language_code in FAIRSEQ_LANGUAGE_CODES.iter() {
+            let language_code = if language_code.len() == 3 {
+                format!(">>{language_code}<<")
+            } else {
+                return VocabularyValidationSnafu {
+                    message: "NLLB Vocab only supports language code of length 8",
+                }
+                .fail();
+            };
+            values.insert(language_code.clone(), values.len() as i64);
+            NLLBVocab::_register_as_special_value(
+                language_code.as_str(),
+                values,
+                &mut special_values,
+            )?;
+        }
+
+        // TODO: remove it (it's already contained in `added_tokens`):
+        let vocab = &tokenizer.model.vocab;
+
+        let indices = swap_key_values(vocab);
+        let special_indices = swap_key_values(&special_values);
+        let language_codes_bytes = FAIRSEQ_LANGUAGE_CODES
+            .iter()
+            .map(|f| format!(">>{f}<<").as_bytes().to_vec())
+            .collect::<HashSet<Vec<u8>>>();
+
+        Ok(Self {
+            indices,
+            language_codes_bytes,
+            special_indices,
+            special_values,
+            values: tokenizer.model.vocab,
+            special_token_storage,
+        })
+    }
+
+    fn token_to_id(&self, token: &str) -> i64 {
+        self._token_to_id(
+            token,
+            self.values(),
+            self.special_values(),
+            self.get_unknown_value(),
+        )
+    }
+
+    fn id_to_token(&self, id: &i64) -> String {
+        self._id_to_token(
+            id,
+            self.indices(),
+            self.special_indices(),
+            self.get_unknown_value(),
+        )
+    }
+}
+
+#[derive(Deserialize)]
+#[allow(dead_code)]
```

Comment:
Where are these used (is the `#[allow(dead_code)]` required?)

Reply:
It is used in `NLLBVocab`, I'll remove this modifier later.

Reply:
Fixed in #81.